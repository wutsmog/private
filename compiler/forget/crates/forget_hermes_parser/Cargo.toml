[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }