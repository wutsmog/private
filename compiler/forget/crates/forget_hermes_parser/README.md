# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.