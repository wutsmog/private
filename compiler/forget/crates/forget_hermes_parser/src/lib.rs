mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
