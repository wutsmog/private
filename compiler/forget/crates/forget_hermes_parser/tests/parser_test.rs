use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
