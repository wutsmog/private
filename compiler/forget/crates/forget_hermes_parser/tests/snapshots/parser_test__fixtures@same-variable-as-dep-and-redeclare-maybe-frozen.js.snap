---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/same-variable-as-dep-and-redeclare-maybe-frozen.js
---
Input:
// note: comments are for the ideal scopes, not what is currently
// emitted
function foo(props) {
  // scope 0: deps=[props.a] decl=[x] reassign=none
  let x = [];
  x.push(props.a);

  // scope 1: deps=[x] decl=[header] reassign=none
  const header = props.showHeader ? <div>{x}</div> : null;

  // scope 2:
  // deps=[x, props.b, props.c]
  // decl=none
  // reassign=[x]
  const y = [x]; // y depends on the earlier x
  x = []; // x reassigned
  y.push(props.b); // interleaved mutation of x/y
  x.push(props.c); // interleaved mutation

  // scope 3 ...
  const content = (
    <div>
      {x}
      {y}
    </div>
  );

  // scope 4 ...
  return (
    <>
      {header}
      {content}
    </>
  );
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "foo",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 86,
          "end": 89
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 90,
            "end": 95
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "let",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 157,
                    "end": 158
                  }
                },
                "init": {
                  "type": "ArrayExpression",
                  "elements": [],
                  "loc": null,
                  "range": {
                    "start": 161,
                    "end": 163
                  }
                },
                "loc": null,
                "range": {
                  "start": 157,
                  "end": 163
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 153,
              "end": 164
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 167,
                    "end": 168
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "push",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 169,
                    "end": 173
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 167,
                  "end": 173
                }
              },
              "arguments": [
                {
                  "type": "MemberExpression",
                  "object": {
                    "type": "Identifier",
                    "name": "props",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 174,
                      "end": 179
                    }
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "a",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 180,
                      "end": 181
                    }
                  },
                  "computed": false,
                  "loc": null,
                  "range": {
                    "start": 174,
                    "end": 181
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 167,
                "end": 182
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 167,
              "end": 183
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "header",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 244,
                    "end": 250
                  }
                },
                "init": {
                  "type": "ConditionalExpression",
                  "test": {
                    "type": "MemberExpression",
                    "object": {
                      "type": "Identifier",
                      "name": "props",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 253,
                        "end": 258
                      }
                    },
                    "property": {
                      "type": "Identifier",
                      "name": "showHeader",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 259,
                        "end": 269
                      }
                    },
                    "computed": false,
                    "loc": null,
                    "range": {
                      "start": 253,
                      "end": 269
                    }
                  },
                  "alternate": {
                    "type": "NullLiteral",
                    "loc": null,
                    "range": {
                      "start": 289,
                      "end": 293
                    }
                  },
                  "consequent": {
                    "type": "JSXElement",
                    "openingElement": {
                      "type": "JSXOpeningElement",
                      "name": {
                        "type": "JSXIdentifier",
                        "name": "div",
                        "loc": null,
                        "range": {
                          "start": 273,
                          "end": 276
                        }
                      },
                      "attributes": [],
                      "selfClosing": false,
                      "loc": null,
                      "range": {
                        "start": 272,
                        "end": 277
                      }
                    },
                    "children": [
                      {
                        "type": "JSXExpressionContainer",
                        "expression": {
                          "type": "Identifier",
                          "name": "x",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 278,
                            "end": 279
                          }
                        },
                        "loc": null,
                        "range": {
                          "start": 277,
                          "end": 280
                        }
                      }
                    ],
                    "closingElement": {
                      "type": "JSXClosingElement",
                      "name": {
                        "type": "JSXIdentifier",
                        "name": "div",
                        "loc": null,
                        "range": {
                          "start": 282,
                          "end": 285
                        }
                      },
                      "loc": null,
                      "range": {
                        "start": 280,
                        "end": 286
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 272,
                      "end": 286
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 253,
                    "end": 293
                  }
                },
                "loc": null,
                "range": {
                  "start": 244,
                  "end": 293
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 238,
              "end": 294
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 383,
                    "end": 384
                  }
                },
                "init": {
                  "type": "ArrayExpression",
                  "elements": [
                    {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 388,
                        "end": 389
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 387,
                    "end": 390
                  }
                },
                "loc": null,
                "range": {
                  "start": 383,
                  "end": 390
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 377,
              "end": 391
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "AssignmentExpression",
              "operator": "=",
              "left": {
                "type": "Identifier",
                "name": "x",
                "typeAnnotation": null,
                "loc": null,
                "range": {
                  "start": 424,
                  "end": 425
                }
              },
              "right": {
                "type": "ArrayExpression",
                "elements": [],
                "loc": null,
                "range": {
                  "start": 428,
                  "end": 430
                }
              },
              "loc": null,
              "range": {
                "start": 424,
                "end": 430
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 424,
              "end": 431
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 450,
                    "end": 451
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "push",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 452,
                    "end": 456
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 450,
                  "end": 456
                }
              },
              "arguments": [
                {
                  "type": "MemberExpression",
                  "object": {
                    "type": "Identifier",
                    "name": "props",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 457,
                      "end": 462
                    }
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "b",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 463,
                      "end": 464
                    }
                  },
                  "computed": false,
                  "loc": null,
                  "range": {
                    "start": 457,
                    "end": 464
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 450,
                "end": 465
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 450,
              "end": 466
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 500,
                    "end": 501
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "push",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 502,
                    "end": 506
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 500,
                  "end": 506
                }
              },
              "arguments": [
                {
                  "type": "MemberExpression",
                  "object": {
                    "type": "Identifier",
                    "name": "props",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 507,
                      "end": 512
                    }
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "c",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 513,
                      "end": 514
                    }
                  },
                  "computed": false,
                  "loc": null,
                  "range": {
                    "start": 507,
                    "end": 514
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 500,
                "end": 515
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 500,
              "end": 516
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "content",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 567,
                    "end": 574
                  }
                },
                "init": {
                  "type": "JSXElement",
                  "openingElement": {
                    "type": "JSXOpeningElement",
                    "name": {
                      "type": "JSXIdentifier",
                      "name": "div",
                      "loc": null,
                      "range": {
                        "start": 584,
                        "end": 587
                      }
                    },
                    "attributes": [],
                    "selfClosing": false,
                    "loc": null,
                    "range": {
                      "start": 583,
                      "end": 588
                    }
                  },
                  "children": [
                    {
                      "type": "JSXText",
                      "value": "\n      ",
                      "raw": "\n      ",
                      "loc": null,
                      "range": {
                        "start": 588,
                        "end": 595
                      }
                    },
                    {
                      "type": "JSXExpressionContainer",
                      "expression": {
                        "type": "Identifier",
                        "name": "x",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 596,
                          "end": 597
                        }
                      },
                      "loc": null,
                      "range": {
                        "start": 595,
                        "end": 598
                      }
                    },
                    {
                      "type": "JSXText",
                      "value": "\n      ",
                      "raw": "\n      ",
                      "loc": null,
                      "range": {
                        "start": 598,
                        "end": 605
                      }
                    },
                    {
                      "type": "JSXExpressionContainer",
                      "expression": {
                        "type": "Identifier",
                        "name": "y",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 606,
                          "end": 607
                        }
                      },
                      "loc": null,
                      "range": {
                        "start": 605,
                        "end": 608
                      }
                    },
                    {
                      "type": "JSXText",
                      "value": "\n    ",
                      "raw": "\n    ",
                      "loc": null,
                      "range": {
                        "start": 608,
                        "end": 613
                      }
                    }
                  ],
                  "closingElement": {
                    "type": "JSXClosingElement",
                    "name": {
                      "type": "JSXIdentifier",
                      "name": "div",
                      "loc": null,
                      "range": {
                        "start": 615,
                        "end": 618
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 613,
                      "end": 619
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 583,
                    "end": 619
                  }
                },
                "loc": null,
                "range": {
                  "start": 567,
                  "end": 623
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 561,
              "end": 624
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "JSXFragment",
              "openingFragment": {
                "type": "JSXOpeningFragment",
                "loc": null,
                "range": {
                  "start": 658,
                  "end": 660
                }
              },
              "children": [
                {
                  "type": "JSXText",
                  "value": "\n      ",
                  "raw": "\n      ",
                  "loc": null,
                  "range": {
                    "start": 660,
                    "end": 667
                  }
                },
                {
                  "type": "JSXExpressionContainer",
                  "expression": {
                    "type": "Identifier",
                    "name": "header",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 668,
                      "end": 674
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 667,
                    "end": 675
                  }
                },
                {
                  "type": "JSXText",
                  "value": "\n      ",
                  "raw": "\n      ",
                  "loc": null,
                  "range": {
                    "start": 675,
                    "end": 682
                  }
                },
                {
                  "type": "JSXExpressionContainer",
                  "expression": {
                    "type": "Identifier",
                    "name": "content",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 683,
                      "end": 690
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 682,
                    "end": 691
                  }
                },
                {
                  "type": "JSXText",
                  "value": "\n    ",
                  "raw": "\n    ",
                  "loc": null,
                  "range": {
                    "start": 691,
                    "end": 696
                  }
                }
              ],
              "closingFragment": {
                "type": "JSXClosingFragment",
                "loc": null,
                "range": {
                  "start": 696,
                  "end": 699
                }
              },
              "loc": null,
              "range": {
                "start": 658,
                "end": 699
              }
            },
            "loc": null,
            "range": {
              "start": 645,
              "end": 704
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 97,
          "end": 706
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 77,
        "end": 706
      },
      "loc": null,
      "range": {
        "start": 77,
        "end": 706
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 77,
    "end": 706
  }
}
